internal/etcd/etcd.go

```diff
@@ -149,6 +151,74 @@ func (c *ClientEtcdAPI) Close() error {
 	return nil
 }
 
+// UsedSpacePercentage returns the used space as a percentage of the storage quota
+func (c *ClientEtcdAPI) UsedSpacePercentage(ctx context.Context, member Member) (int, error) {
+	const (
+		storageQuotaKey = "etcd_server_quota_backend_bytes"
+		storageSizeKey  = "etcd_mvcc_db_total_size_in_bytes"
+	)
+	if len(member.ClientURLs) < 1 {
+		return 0, fmt.Errorf("unable to get used space, etcd member has no client urls")
+	}
+	url := member.ClientURLs[0] + "/metrics"
+	resp, err := http.Get(url)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get metrics from etcd: %w", err)
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read metrics returned from etcd: %w", err)
+	}
+	defer resp.Body.Close()
+
+	strs := strings.Split(string(b), "\n")
+	var metricsWeCareAbout []string
+	for _, str := range strs {
+		if !strings.Contains(str, "HELP") &&
+			!strings.Contains(str, "TYPE") &&
+			(strings.Contains(str, storageQuotaKey) ||
+				strings.Contains(str, storageSizeKey)) {
+			metricsWeCareAbout = append(metricsWeCareAbout, str)
+		}
+	}
+
+	var storageQuota float64
+	var storageSize float64
+
+	for _, metric := range metricsWeCareAbout {
+		str := strings.Split(metric, " ")
+		switch str[0] {
+		case storageQuotaKey:
+			storageQuota, err = strconv.ParseFloat(str[1], 32)
+			if err != nil {
+				return 0, fmt.Errorf("failed to convert storage quota to float: %w", err)
+			}
+		case storageSizeKey:
+			storageSize, err = strconv.ParseFloat(str[1], 32)
+			if err != nil {
+				return 0, fmt.Errorf("failed to convert storage size to int: %w", err)
+			}
+		}
+	}
+	fmt.Printf("storage size %v     storage quota %v\n", storageSize, storageQuota)
+
+	usedSpace := storageSize / storageQuota * 100.0
+	return int(usedSpace), nil
+}
+
+func (c *ClientEtcdAPI) Defragment(ctx context.Context, member Member) error {
+	if c.ClientV3 != nil {
+		if len(member.ClientURLs) < 1 {
+			return errors.New("member needs at least 1 client url to be defragged")
```

Comment:
I think early returns should move business logic to the left.

Reply:
Yep, done :)